lab/lablib/tone_helpers/compile_music.js

```diff
@@ -95,30 +263,88 @@ export function compile_sequence(tone, instrument, midi_cycle) {
   // TODO: Need to compute this, but determining how to pass this to Cycle
   // would require something other than an array
   const note_duration = "16n";
-  return new tone.Sequence(
+  const seq = new tone.Sequence(
     (time, pitch) => {
       instrument.triggerAttackRelease(pitch, note_duration, time);
     },
     pattern,
     tone_interval
   );
+  seq.loop = true;
+
+  return new ToneClip(seq, midi_cycle.duration);
+}
+
+/**
+ * @param {import("tone")} tone the Tone.js library
+ * @param {import("tone").Synth} instrument the instrument to play
+ * @param {MusicLoop<number>} midi_loop Loop of MIDI pitches
+ * @returns {Loop<ToneClip>} A sequence of generated clips, or undefined if there were no notes
+ */
+export function compile_loop(tone, instrument, midi_loop) {
+  const child = compile_music(tone, instrument, midi_loop.child);
+  return new Loop(midi_loop.duration, child);
+}
+
+export class ToneClip {
+  /**
+   * Constructor
+   * @param {import("tone").ToneEvent} material The
+   * @param {Rational} duration The duration of the clip in measures
+   */
+  constructor(material, duration) {
+    this.material = material;
+    this.duration = duration;
+  }
 }
 
 /**
  * Compile music to a Tone.js Part or Sequence
  * @param {import("tone")} tone The Tone.js library
  * @param {import("tone").Synth} instrument
  * @param {import("../music/Score.js").Music<number>} music
- * @return {import("tone").Sequence | import("tone").Part} The compiled music
+ * @return {import("../music/Timeline.js").Timeline<ToneClip>} The compiled music clips, or undefined if the timeline is empty
  */
 export function compile_music(tone, instrument, music) {
```

Comment:
To make unit testing possible, split this into `precompile_music() -> Timeline<ClipDescriptor>` which creates arrays of Tone.js-style strings for pitch and duration (as this is unit testable). Then have a `timeline_map` to compile these clips.